torchgeo/datasets/substation_seg.py

```diff
@@ -0,0 +1,196 @@
+
+        Args:
+            index: Index of the item to retrieve.
+
+        Returns:
+            A dictionary containing the image and corresponding mask.
+        """
+        image_filename = self.image_filenames[index]
+        image_path = os.path.join(self.image_dir, image_filename)
+        mask_path = os.path.join(self.mask_dir, image_filename)
+
+        image = np.load(image_path)['arr_0']
+        mask = np.load(mask_path)['arr_0']
+
+        # Standardize images
+        image = self._normalize_image(image)
+
+        # Handle multiple channels and timepoints
+        image = self._handle_channels_and_timepoints(image)
+
+        # Process mask
+        mask[mask != 3] = 0
+        mask[mask == 3] = 1
+        mask = torch.from_numpy(mask).float().unsqueeze(dim=0)
+
+        # Apply geo and color transformations
+        image, mask = self._apply_transforms(image, mask)
+
+        return {"image": image, "mask": mask}
+
+    def __len__(self) -> int:
+        """Return the length of the dataset.
+
+        Returns:
+            The number of items in the dataset.
+        """
+        return len(self.image_filenames)
+
+    def plot(self) -> None:
+        """Plot a random image and mask from the dataset."""
+        index = np.random.randint(0, self.__len__())
+        data = self.__getitem__(index)
+        image, mask = data["image"], data["mask"]
+        fig, axs = plt.subplots(1, 2, figsize=(15, 15))
+        axs[0].imshow(image.permute(1, 2, 0))
+        axs[1].imshow(image.permute(1, 2, 0))
+        axs[1].imshow(mask.permute(1, 2, 0), alpha=0.5, cmap='gray')
+
+    def _normalize_image(self, image: np.ndarray[Any, np.float64]) -> torch.Tensor:
+        """Normalize the image based on the selected normalizing type.
+
+        Args:
+            image: The image to normalize.
+
+        Returns:
+            Normalized image as a torch tensor.
+        """
+        if self.normalizing_type == 'percentile':
+            image = (image - self.normalizing_factor[:, 0].reshape((-1, 1, 1))) / self.normalizing_factor[:, 2].reshape((-1, 1, 1))
+        elif self.normalizing_type == 'zscore':
+            image = (image - self.args.means) / self.args.stds
+        else:
+            image = image / self.normalizing_factor
+            image = np.clip(image, 0, 1)
+        return torch.from_numpy(image)
+
+    def _handle_channels_and_timepoints(self, image: np.ndarray[Any, np.float64]) -> torch.Tensor:
+        """Handle channels and timepoints in the image.
+
+        Args:
+            image: The image to process.
+
+        Returns:
+            Processed image as a torch tensor.
+        """
+        if self.in_channels == 3:
+            image = image[:, [3, 2, 1], :, :]
+        else:
+            image = image[:4, :, :, :] if self.use_timepoints else image[0]
+        return torch.from_numpy(image)
+
+    def _apply_transforms(self, image: torch.Tensor, mask: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
```

Comment:
Hi @rijuld, thank you for contributing this dataset. As another pointer, torchgeo datasets usually have an accompanying datamodule that defines things like the train/val/test split, but also common data augmentations, like flips, color augmentations etc through the kornia package. So in essence, torchgeo datasets simply load a particular sample and the augmentations are applied on GPU over the batch. 

Let me know if I can help with any further questions.

For example in this dataset, the `getitem` [method](https://github.com/zsyOAOA/ResShift/blob/2bc69fdd39827385c91db49cb1d155c69d5e85e8/trainer.py#L533) loads the image and mask, and then we have a corresponding [datamodule](https://github.com/microsoft/torchgeo/blob/f2ed33a4079f632a107096df8efdebe711a0bb91/torchgeo/datamodules/caffe.py#L43) where we define augmentations like resizing and others, which will automatically be applied with a lightning training setup. This helps streamlining the datasets and keep them "minimal" and also make use of existing augmentation implementations like Kornia.

Reply:
Hi @nilsleh , thank you for the detailed explanation!

That makes perfect sense. I will try to make this minimal, implement this today and reach out if I have any further questions.

Thanks again!

Reply:
Apologies for the late response. Adam prefers having all data normalization in the datamodule for consistency, But I also don't think it is terrible to do in the dataset. If you move it to the datamodule, you can use the kornina `Normalize` module, that you can add to the augmentation series. Then it will be applied to `on_after_batch_transfer` in the `LightningDataModule`.